app/controllers/WhatDoYouWantToDoController.scala

```diff
@@ -120,12 +118,18 @@ class WhatDoYouWantToDoController @Inject()(
 
   private def allowWhatDoYouWantToDo(implicit request: Request[AnyContent], user: AuthedUser): Future[Result] = {
     val nino = user.nino
-    for {
-      hasTaxCodeChanged <- taxCodeChangeService.hasTaxCodeChanged(nino)
-      showJrsTile       <- jrsService.checkIfJrsClaimsDataExist(nino)
-      model             <- whatToDoView(nino, hasTaxCodeChanged, showJrsTile)
-      _                 <- auditNumberOfTaxCodesReturned(nino, showJrsTile)
-    } yield Ok(whatDoYouWantToDoTileView(WhatDoYouWantToDoForm.createForm, model, applicationConfig))
+
+    taxCodeChangeService.hasTaxCodeChanged(nino) flatMap {
+      case Right(taxCodeChanged) =>
+        for {
+          showJrsTile <- jrsService.checkIfJrsClaimsDataExist(nino)
+          model       <- whatToDoView(nino, taxCodeChanged, showJrsTile)
+          _           <- auditNumberOfTaxCodesReturned(nino, showJrsTile)
+        } yield Ok(whatDoYouWantToDoTileView(WhatDoYouWantToDoForm.createForm, model, applicationConfig))
+      case Left(taxCodeError) =>
+        Future.successful(
+          BadRequest(errorPagesHandler.error4xxPageWithLink(taxCodeError.errorMessage.getOrElse("Tax Code not Found"))))
```

Comment:
I don't think the expected behaviour is a BadRequest as the previous code used: 
```final case class TaxCodeError(nino: Nino, errorMessage: Option[String] = None) extends Error``` 
which is not a subset of ```BadRequestException``` 

In order to maintain the same behaviour this should be used: 
```
errorPagesHandler.hodBadRequestResult(nino.toString()) orElse errorPagesHandler.hodInternalErrorResult(
        nino.toString())
        ```

Reply:
I would propagate the Future[Either[_,_]] in the allowWhatDoYouWantToDo and then use the errorPagesHandler

Reply:
Ok 👍 Will incorporate that change. Thanks